`RemovePallet` migration tool

liamaharon opened the pull request:
- Creates `RemovePallet` which can be used to remove all storage items for a given pallet on the next runtime upgrade.
- ~Uses `RemovePallet` to delete storage for Kusama pallets that were removed from the runtime in https://github.com/paritytech/polkadot/pull/6701/~

Related: https://github.com/paritytech/polkadot/issues/6749

liamaharon commented on runtime/kusama/src/lib.rs:
```diff
@@ -1326,6 +1327,97 @@ impl pallet_nomination_pools::Config for Runtime {
+			let estimated_weight = batch.get_dispatch_info().weight;
+			match RuntimeCall::Utility(batch).dispatch(RuntimeOrigin::root()) {
+				Ok(post_info) => post_info.actual_weight.unwrap_or_default(),
+				Err(_) => estimated_weight,
```
I wasn't exactly sure what to return for weight in this arm. Let me know if there's something else I should do.

kianenigma commented on runtime/kusama/src/lib.rs:
```diff
@@ -1326,6 +1327,97 @@ impl pallet_nomination_pools::Config for Runtime {
 
+pub struct RemoveGovV1Storage;
+impl RemoveGovV1Storage {
+	fn get_pallets() -> Vec<&'static str> {
```
nit, but a bit superfluous. FYI a struct can also have associated `const` items. A `const MODULES: [&'static str; 6] = ..` makes more sense.

kianenigma commented on runtime/kusama/src/lib.rs:
```diff
@@ -1326,6 +1327,97 @@ impl pallet_nomination_pools::Config for Runtime {
+			counter += 1;
+			current = next;
+		}
+		0
```
```suggestion
		unreachable!()
```

right? I am not entirely sure.

kianenigma commented on runtime/kusama/src/lib.rs:
```diff
@@ -1326,6 +1327,97 @@ impl pallet_nomination_pools::Config for Runtime {
 	type MaxPointsToBalance = MaxPointsToBalance;
 }
 
+pub struct RemoveGovV1Storage;
```
Could/should go into `runtime-common` crate to be usable in more places.

kianenigma commented on runtime/kusama/src/lib.rs:
```diff
@@ -1326,6 +1327,97 @@ impl pallet_nomination_pools::Config for Runtime {
+			.iter()
+			.map(|pallet| {
+				let prefix = twox_128(pallet.as_bytes());
+				let subkeys = Self::count_keys(&prefix);
```
I would add a new helper similar to https://paritytech.github.io/substrate/master/frame_support/storage/migration/fn.move_pallet.html that is `remove_pallet(prefix: &'static str) -> u32` that removes and counts at the same time. 

In your current code, you are doing the expensive iteration over all keys twice, once for counting and once for deleting.

liamaharon commented:
Question: since we're moving this to `runtime-common`, does it make sense to make the implementation applicable to any set of modules rather than being hard coded to only support Gov V1? 

e.g. rather than
```rust
pub type Migrations = (
	RemoveGovV1Storage
);
```

you'd use it something like 

```rust
pub type Migrations = (
	RemoveModuleStorage("Module1", "Module2", "Module3", ...)
);
```

kianenigma commented:
> Question: since we're moving this to `runtime-common`, does it make sense to make the implementation applicable to any set of modules rather than being hard coded to only support Gov V1?
> 
> e.g. rather than
> 
> ```rust
> pub type Migrations = (
> 	RemoveGovV1Storage
> );
> ```
> 
> you'd use it something like
> 
> ```rust
> pub type Migrations = (
> 	RemoveModuleStorage("Module1", "Module2", "Module3", ...)
> );
> ```

Yes, you can make something like:

```rust 
struct RemovePallet<P: Get<&'static str>>;
// or 
struct RemovePallets<P: Get<Vec<&'static str>>>;
```

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2607718

liamaharon commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,81 @@
+			current = next;
+		}
+		// Extra read for initial prefix read.
+		(counter, RocksDbWeight::get().reads_writes(counter + 1, counter))
```
Could someone please check this weight calculation?

liamaharon commented:
@kianenigma I've refactored the code so it supports removing keys for arbitrary pallets and moved it to live under `runtime/common`.

When you get some time I'd appreciate another review.

bkchr commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,81 @@
+}
+impl<P: Get<&'static str>> frame_support::traits::OnRuntimeUpgrade for RemovePallet<P> {
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		Self::clear_keys(false).1
```
Could be https://github.com/paritytech/substrate/blob/5b64a4090cb02c47db31310cb2120da6ede6452e/frame/support/src/storage/unhashed.rs#L140-L144

bkchr commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,81 @@
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+		let (key_count, _) = Self::clear_keys(true);
+		if key_count > 0 {
+			log::info!("Found {} keys for pallet {} pre-removal 👀", key_count, P::get());
+		} else {
+			log::warn!("No keys found for pallet {} pre-removal ⚠️", P::get());
+		}
+		Ok(Vec::new())
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(_state: Vec<u8>) -> Result<(), &'static str> {
+		let (key_count, _) = Self::clear_keys(true);
+		if key_count > 0 {
+			log::error!("{} pallet {} keys remaining post-removal ❗", key_count, P::get());
+		} else {
+			log::info!("No {} keys remaining post-removal 🎉", P::get())
+		}
+		Ok(())
+	}
```
And these could be https://github.com/paritytech/substrate/blob/5b64a4090cb02c47db31310cb2120da6ede6452e/frame/support/src/storage/unhashed.rs#L160-L165

liamaharon commented:
Thanks @bkchr, much cleaner now 👌

bkchr commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,68 @@
+		let hashed_prefix = twox_128(P::get().as_bytes());
+		let keys_removed = match clear_prefix(&hashed_prefix, None) {
+			KillStorageResult::AllRemoved(value) => value,
+			KillStorageResult::SomeRemaining(value) => value,
```
This should not happen, but yeah we hopefully catch this with `post_upgrade`!

ggwpez commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,68 @@
+use sp_io::{hashing::twox_128, storage::clear_prefix, KillStorageResult};
+use sp_std::{marker::PhantomData, vec::Vec};
+
+pub struct RemovePallet<P: Get<&'static str>>(PhantomData<P>);
```
Should the doc not rather be on this struct?  
And please doc how to use the `P`.

ggwpez commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,68 @@
+		let hashed_prefix = twox_128(P::get().as_bytes());
+		let keys_removed = match clear_prefix(&hashed_prefix, None) {
+			KillStorageResult::AllRemoved(value) => value,
+			KillStorageResult::SomeRemaining(value) => value,
```
We could do a storage roll-back in this case, but it is probably fine since the pallets should not be used anyway.

ggwpez replied:
But please add an error log, just in case. The post_upgrade does not run for the real upgrade.

ggwpez commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,68 @@
+//! runtime.
+//!
+//! WARNING: `RemovePallet` has no guard rails preventing it from bricking the chain if the
+//! operation of removing storage for the given pallet would exceed the block weight limit.
```
We could add a simple limit though. The multi-block migrations are not really needed here since this pallet is not in active use anymore.  
Some const generic in the `RemovePallet` for example. But I am also fine with it now, we can extend it later.

ggwpez commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,68 @@
+	fn post_upgrade(_state: Vec<u8>) -> Result<(), &'static str> {
+		let hashed_prefix = twox_128(P::get().as_bytes());
+		match contains_prefixed_key(&hashed_prefix) {
+			true => log::error!("{} has keys remaining post-removal ❗", P::get()),
```
This case should return an error.

ggwpez commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,68 @@
+		let hashed_prefix = twox_128(P::get().as_bytes());
+		match contains_prefixed_key(&hashed_prefix) {
+			true => log::info!("Found {} keys pre-removal 👀", P::get()),
+			false => log::warn!("No {} keys found pre-removal ⚠️", P::get()),
```
```suggestion
			false => log::warn!("Migration RemovePallet<{}> can be removed (no keys found pre-removal).", P::get()),
```
It is good to think about how someone who has zero idea about what is going on will interpret the log output. So keeping it very self-explaining is an advantage.

ggwpez reviewed (commented):
Good idea!

ggwpez commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,68 @@
+		let hashed_prefix = twox_128(P::get().as_bytes());
+		match contains_prefixed_key(&hashed_prefix) {
+			true => log::error!("{} has keys remaining post-removal ❗", P::get()),
+			false => log::info!("No {} keys found post-removal 🎉", P::get()),
```
You could also calculate the difference pre- and post to see how many keys got removed. The `Vec` in the pre- and post can be used to pass encoded data.

bkchr replied:
Not sure we need this and `on_runtime_upgrade` already prints the number.

liamaharon replied:
> Not sure we need this and `on_runtime_upgrade` already prints the number.

I was thinking this, but happy to add it to `post` anyway @ggwpez if you prefer.

ggwpez reviewed (approved):
Thanks!

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2628031

xlc commented:
how about locked/reserved funds?

liamaharon commented:
> how about locked/reserved funds?

This tool is specifically for removing storage. Something for migrating/recovering locked/reserved funds I think should be part of a seperate work effort. Sounds like deserving of a new issue to me.

kianenigma commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,125 @@
+
+		log::info!("Removed {} {} keys 🧹", keys_removed, P::get());
+
+		RocksDbWeight::get().reads_writes(keys_removed + 1, keys_removed)
```
```suggestion
		DbWeights::get().reads_writes(keys_removed + 1, keys_removed)
```

kianenigma commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -0,0 +1,125 @@
+/// items (and performing other heavy migrations) over multiple blocks.
+/// (https://github.com/paritytech/substrate/issues/13690)
+pub struct RemovePallet<P: Get<&'static str>>(PhantomData<P>);
+impl<P: Get<&'static str>> frame_support::traits::OnRuntimeUpgrade for RemovePallet<P> {
```
```suggestion
impl<P: Get<&'static str>, DbWeight: Get<DbWeight>> frame_support::traits::OnRuntimeUpgrade for RemovePallet<P> {
```

To support anyone not using RocksDB in principle.

kianenigma commented:
> how about locked/reserved funds?

I thought the PR that kickstarted Gov V2 already did that. 

If not, it will be much harder to do once we remove the storage. We should also do all the unreserve/unlocks in this PR.

xlc commented:
We have not yet disabled gov1 and the migration must happen after it is disabled before it is removed

ggwpez reviewed (changes requested):
Yea going to block so we dont accidentally merge it until gov1 is removed.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2645642

ggwpez commented on runtime/common/src/remove_pallet.rs:
```diff
@@ -77,8 +84,12 @@ use sp_std::{marker::PhantomData, vec::Vec};
 /// (https://github.com/paritytech/substrate/issues/13690)
-pub struct RemovePallet<P: Get<&'static str>>(PhantomData<P>);
-impl<P: Get<&'static str>, DbWeight: Get<DbWeight>> frame_support::traits::OnRuntimeUpgrade for RemovePallet<P> {
+pub struct RemovePallet<P: Get<&'static str>, DbWeight: Get<RuntimeDbWeight>>(
```
We could also move it to Substrate… :man_shrugging:  
Not sure if it is worth it right now though.

bkchr replied:
Probably `frame-support`.
